### PR TITLE
Auto-build releases (#282)

### PR DESCRIPTION
**Pull request by artoonie**
Github action to:
1. Build assets on every merge into develop and master
2. Automatically add assets to every prerelease and release

Where assets include the zip files for windows, mac, and linux, built with jdk 17.0.2, along with their SHA512s.

The zips go in two places:
1. On every push to master or develop, they will show up [in the "artifacts" section](https://github.com/BrightSpots/rcv/actions/runs/4773447699) in the Action linked from the PR.
2. On every prerelease and release, they will show up [in the Releases section](https://github.com/BrightSpots/rcv/releases)

### CODE REVIEW

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:5]: This means it'll automatically trigger when we save a draft of a new release? Or do we specifically have to choose the "pre-release" box?

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:8]: How does it know to build the v1.3.0 or whatever tag in master?

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:9]: What is develop doing in here? Sorry I was lazy and didn't look up the docs around this...

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:35]: Where'd you pull this hash from? Is there a list somewhere to choose from?

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:29]: When we update the JDK, is it as simple as incrementing it here and GH will take care of that? Also, do we ever have to change this `temurin` thing?

File: .github/workflows/release.yml
Comment (HEdingfield) [inline .github/workflows/release.yml:32]: I assume it'll stop the auto-build if this fails? Will we get emailed a notification or how exactly does that work?

**Review by HEdingfield — COMMENTED** [review]
This is potentially mindblowingly cool! Thanks so much for putting the work in to get this going!

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:5]: See the third info box here: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release

```Note: The prereleased type will not trigger for pre-releases published from draft releases, but the published type will trigger. If you want a workflow to run when stable and pre-releases publish, subscribe to published instead of released and prereleased.```

So this will trigger when you draft a release or a prerelease, but not a draft. That's intentional, because drafts would have access to all the same information via the branch tags, just not uploaded to the "Release" section.

I'll clarify this in the main PR comment.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:8]: 1. If the trigger is a push to master or develop, the tag will be the branch name
2. If the trigger is a release, the tag will be the tag name, which is equal to the version number

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:9]: See the updated main comment above. This is Option (1), creating Assets, which won't go into the Releases section.

We can remove this, but it's designed to allow non-devs to quickly test the latest Develop or Master without needing to compile code.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:35]: From the Getting Started demos from github: https://github.com/actions/starter-workflows/blob/main/ci/gradle-publish.yml

The list would be any commit here: https://github.com/gradle/gradle-build-action/commits/main

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:29]: 1. Yep, we just increment it here
2. Not sure about temurin

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:32]: 1. The release creator will get an email notification if it failed while creating a release
2. It will mark the branch as having a failed build + we'll get an email notification if it happens after a merge to master/develop, like so:
 
image

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:9]: I think it's a cool idea, but let's remove it for now and add it back in in the future if there's ever any demonstrated need for it.

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:5]: Cool, ok. The only thing I'm confused about is how you supposedly created a [`test-auto-release`](https://github.com/BrightSpots/rcv/releases/tag/test-auto-release) tag, but I don't see it in any of the branches when I run `git tag`. Know why this is? Does it hold off on pushing the tag until the non-pre-release is published?

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:29]: Apparently temurin is an anagram of "runtime", and it's associated with the Eclipse version of the JDK. Interesting. I guess we just keep this as-is.

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:32]: Perfect, that sounds good.

File: .github/workflows/release.yml
Comment (HEdingfield) [reply .github/workflows/release.yml:35]: Interesting, ok. Is this another thing we'd need to update regularly (i.e. before each release?).

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:5]: I created several test tags and then deleted them, so `test-auto-release` was a tag I created and deleted while debugging the workflow

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:9]: If I could add another reason to keep this: (1) it runs faster than the tests, and runs in parallel, so there are no downsides, and (2) it allows regular testing of this workflow, instead of only running the workflow on releases.

I'm a little afraid it will become stale if we only run it on release creation.

If you still think I should remove it, I'm happy to.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:35]: I don't think so -- perhaps if gradle has major security updates?

We can also have it use the latest commit, but locking the version is more secure and more reliable.

File: .github/workflows/release.yml
Comment (artoonie) [reply .github/workflows/release.yml:9]: Some thoughts:
1. I could skip the artifact upload if that's an issue
2. Or, decrease the lifespan to 5 days

But, for public repos, [my understanding](https://github.com/orgs/community/discussions/26438) is that there is no storage limit for Github Actions.

Again, no strong opinions here, but given that it's free in terms of both money and time, I don't see downsides.

Comment (tarheel) [reply .github/workflows/release.yml:9]: If it's totally free, sure.

Comment (HEdingfield) [reply .github/workflows/release.yml]: Appreciate that, @artoonie. Sounds good re: the draft method.

@tarheel haha I knew that was coming! Probably relatively low. In all seriousness though, my main motivation for working on this project is the hope that it will one day unblock solutions to the climate crisis. I was also brain damaged by doing lots of LCA work in the past, so these are the things that keep me up at night.

**Review by HEdingfield — APPROVED** [review]
Awesome job on this, thanks so much for all the research that went into it!

**Pull request merged**
